azkaban-common/src/main/java/azkaban/executor/container/watch/FlowStatusUpdatingListener.java

```diff
@@ -0,0 +1,317 @@
+import java.util.function.Function;
+import javax.inject.Inject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Provides callback implementations of {@link AzPodStatusListener} for
+ * (1) Updating WebServer and database state of flows based on Pod status
+ * (2) Driving pod lifecycle actions, such as deleting flow pods in a final state.
+ */
+public class FlowStatusUpdatingListener implements AzPodStatusListener {
+
+  private static final Logger logger = LoggerFactory.getLogger(FlowStatusUpdatingListener.class);
+  public static final int EVENT_CACHE_STATS_FREQUENCY = 100;
+  public static final int DEFAULT_EVENT_CACHE_MAX_ENTRIES = 4096;
+
+  private final ContainerizedImpl containerizedImpl;
+  private final ExecutorLoader executorLoader;
+  private final AlerterHolder alerterHolder;
+  private final Cache<String, AzPodStatusMetadata> podStatusCache;
+
+  // Note about the cache size.
+  // Each incoming event is expected to be less than 5KB in size and the maximum cache size will be
+  // about 5KB * maxCacheEntries.
+  private final int maxCacheEntries;
+
+  // Convenience member for referring to the Cache through ConcurrentMap interface.
+  private final ConcurrentMap<String, AzPodStatusMetadata> podStatusMap;
+  private final AtomicLong flowContainerEventCount = new AtomicLong(0);
+
+  @Inject
+  public FlowStatusUpdatingListener(Props azkProps,
+      ContainerizedImpl containerizedImpl,
+      ExecutorLoader executorLoader,
+      AlerterHolder alerterHolder) {
+    requireNonNull(azkProps, "azkaban properties must not be null");
+    requireNonNull(containerizedImpl, "container implementation must not be null");
+    requireNonNull(executorLoader, "executor loader must not be null");
+    requireNonNull(alerterHolder, "alerter holder must not be null");
+    this.containerizedImpl = containerizedImpl;
+    this.executorLoader = executorLoader;
+    this.alerterHolder = alerterHolder;
+
+    maxCacheEntries =
+        azkProps.getInt(ContainerizedDispatchManagerProperties.KUBERNETES_WATCH_DRIVER_THREAD_POOL_SIZE,
+        DEFAULT_EVENT_CACHE_MAX_ENTRIES);
+    this.podStatusCache = CacheBuilder.newBuilder()
+        .maximumSize(maxCacheEntries)
+        .recordStats()
+        .build();
+    this.podStatusMap = podStatusCache.asMap();
+  }
+
+  @VisibleForTesting
+  public ContainerizedImpl getContainerizedImpl() {
+    return this.containerizedImpl;
+  }
+
+  @VisibleForTesting
+  public ExecutorLoader getExecutorLoader() {
+    return this.executorLoader;
+  }
+
+  /**
+   * Validate the the transition from the current state to the new state in the {@code event}
+   * is supported. Note that the validation is on best-effort basis as the cache may not always
+   * contain an event for all running flows. This could happen, for example, during webserver
+   * restarts.
+   * @param event pod watch event
+   */
+  private void validateTransition(AzPodStatusMetadata event) {
```

Comment:
Generally validate methods are expected to return boolean. Can you please do the same here?

Reply:
I have generally seen the convention that methods of the form `is*` return a boolean whereas `validate*` will usually throw exceptions (somewhat like an assertion). In the existing Azkaban code base there are ample examples of `validate*` methods with return types of both void and boolean  (int and other types are also there for that matter). 
Let me know if this is still a concern.

Reply:
Sounds good.